main.c

```diff
@@ -396,46 +446,96 @@ inline uint8_t min(uint8_t a, uint8_t b) { return a <= b ? a : b; }
  * Update lighting table as per animation
  */
 static inline void animationCallback() {
+
   // If the foreground is set we skip the animation as a way to avoid it
   // overrides the foreground
   if (is_foregroundColor_set) {
     return;
   }
 
-  if (profiles[currentProfile].animationSpeed[currentSpeed] > 0) {
-    reactiveNeedsUpdate =
-        profiles[currentProfile].callback(ledColors, ledIntensity);
-
-    if (!reactiveNeedsUpdate) {
-      chBSemReset(&ledDisabledSem, true);
+  bool oldProfileLedActive = profileLedActive;
+  profileLedActive = profiles[currentProfile].callback(ledColors, ledIntensity);
+  if (oldProfileLedActive != profileLedActive) {
+    if (profileLedActive) {
+      enableLeds(false);
+    } else {
+      disableLeds(true);
     }
-  } else {
-    reactiveNeedsUpdate = false;
   }
 }
 
 static inline void sPWM(uint8_t cycle, uint8_t currentCount, ioline_t port) {
   if (cycle > currentCount) {
-    enableRow = true;
     palSetLine(port);
   } else {
     palClearLine(port);
   }
 }
 
-uint32_t animationLastCallTime = 0;
-
 uint8_t rowPWMCount = 0;
 
-THD_WORKING_AREA(waThread1, 128);
-__attribute__((noreturn)) THD_FUNCTION(MsgHandlerThd, arg) {
-  (void)arg;
+// mainCallback is responsible for 2 things:
+// * software PWM
+// * calling animation callback for animated profiles
+void mainCallback(GPTDriver *_driver) {
+  (void)_driver;
 
-  while (true) {
-    msg_t msg;
-    msg = sdGet(&SD1);
-    if (msg >= MSG_OK) {
-      executeMsg(msg);
+  // If LEDs are disabled or the current profile does not need to light up LEDs
+  // (for example with reactive profiles with user inactivity),
+  // this interval will be stopped by a call to disableLeds
+  if (ledEnabled && profileLedActive) {
+
+    palClearLine(ledColumns[currentCol]);
+
+    if (needToCallbackProfile) {
+      needToCallbackProfile = false;
+      profileLedActive =
+          profiles[currentProfile].callback(ledColors, ledIntensity);
+    } else {
+      bool animationCalled = false;
+      if (animationSkipTicks > 0) {
+        // animation update logic
+        uint32_t curTime = chVTGetSystemTimeX();
+        // curTime wraps around when overflows, hence the check for "less"
+        if (curTime < animationLastCallTime ||
+            curTime - animationLastCallTime >= animationSkipTicks) {
+          animationCalled = true;
+          animationCallback();
+          animationLastCallTime = curTime;
+        }
+      }
+      if (!animationCalled) {
```

Comment:
What's the logic behind `animationCalled`?  Seems like that we run the PWM code if there is not animation callback?
I assume the animation doesn't handle PWM. Wouldn't that makes a "always" animating profile keeps animating and never the the pwm update? Or am I understanding something wrong here.

Reply:
Based on my testing, GPT intervals break when I increase the frequency. My guess would be that when the current iteration does not finish in time before the next iteration need to be triggered. Animation callbacks are usually slower that PWM logic. In order to speed up the current iteration, I am doing only one of the things - either PWM or animation. Animation runs much more rare. PWM runs all the time except when animation was run. Reason for combining both was to prevent context switches.

Reply:
Okay we should investigate this a bit more. AFAIK if an IRQ is still in progress, the next IRQ should be queued and not cause any problem. I am kind of busy right now (Just ramped up for my internship), I might not have time to take a look at this in the immediate 2-3 weeks. If you can please do.

Reply:
I will let this go for now but we should look into this.